xrdp/xrdp_mm.c

```diff
@@ -1093,49 +1094,46 @@ dynamic_monitor_data(intptr_t id, int chan_id, char *data, int bytes)
         return 1;
     }
 
-    dynamic_description = (struct dynamic_monitor_description *)
-                          g_malloc(sizeof(struct dynamic_monitor_description), 1);
-    if (!dynamic_description)
+    display_size_data = (struct display_size_description *)
+                        g_malloc(sizeof(struct display_size_description), 1);
+    if (!display_size_data)
     {
         return 1;
     }
-    error = libxrdp_process_monitor_stream(s,
-                                           &(dynamic_description->description),
-                                           1);
+    error = libxrdp_process_monitor_stream(s, display_size_data, 1);
     if (error)
     {
-        LOG(LOG_LEVEL_ERROR, "dynamic_monitor_data: libxrdp_process_monitor_stream"
+        LOG(LOG_LEVEL_ERROR, "dynamic_monitor_data:"
+            " libxrdp_process_monitor_stream"
             " failed with error %d.", error);
+        g_free(display_size_data);
         return error;
     }
-    list_add_item(wm->mm->resize_queue, (tintptr)dynamic_description);
-    dynamic_description->last_state_update_timestamp
-        = dynamic_description->start_time = g_time3();
-    advance_resize_state_machine(
-        wm->mm->resize_state_machine, dynamic_description, WMRZ_QUEUED);
+    list_add_item(wm->mm->resize_queue, (tintptr)display_size_data);
+    g_set_wait_obj(wm->mm->resize_ready);
     return 0;
 }
 
 /******************************************************************************/
 static int
 advance_error(int error,
-              tbus resize_state_machine,
-              struct dynamic_monitor_description *description)
+              struct xrdp_mm *mm,
+              struct display_control_monitor_layout_data *description)
 {
-    advance_resize_state_machine(resize_state_machine, description, WMRZ_ERROR);
+    advance_resize_state_machine(mm, description, WMRZ_ERROR);
     return error;
 }
 
 /******************************************************************************/
 static int
-process_dynamic_monitor_description(struct xrdp_wm *wm,
-                                    struct dynamic_monitor_description *description)
+process_display_control_monitor_layout_data(struct xrdp_wm *wm,
```

Comment:
Thanks for sorting out the state variable.

As there's only one `description` struct in place now, it may be clearer to remove the `description` parameter from this function and also `advance_resize_state_machine()` and `advance_error()`. That will also let you remove the double-indirections in `dynamic_monitor_process_queue()`

Reply:
Updated.